File: templates/graphs/unionFind.cpp
#include <bits/stdc++.h>
using namespace std;
template <typename T>
class UnionFind {
public:
    int Components;
    vector<T> parent, rank;
    UnionFind(int size) : Components(size) {
        parent.resize(size);
        rank.assign(size, 0);
        for (int i = 0; i < size; i++)
            parent[i] = i;
    }
    T find(T x) {
        if (parent[x] != x)
            parent[x] = find(parent[x]);
        return parent[x];
    }
    int size() { return Components; }
    void Union(T a, T b) {
        T aset = find(a), bset = find(b);
        if (aset == bset) return ;
        Components--;
        if (rank[aset] < rank[bset])
            parent[aset] = bset;
        else if (rank[aset] > rank[bset])
            parent[bset] = aset;
        else
            parent[bset] = aset, rank[aset]++;
    }
};
int main(int argc, char const *argv[])
{
    int n, q;
    cin >> n >> q;
    UnionFind<int> obj(n);
    for (int i = 0; i < q; i++)
    {
        int f, a, b;
        cin >> f;
        cin >> a >> b;
        if (f == 1)
            obj.Union(a, b);
        else
            cout << (obj.find(a) == obj.find(b)) << endl;
    }
    return 0;
}

File: python/graphs/UnionFind.py
class UnionFind:
  def __init__(self, n):
    self.rank, self.components = [0] * n, n
    self.parent = list(range(n))

  def find(self, a):
    while self.parent[a] != a:
      self.parent[a] = self.parent[self.parent[a]]
      a = self.parent[a]
    
    return self.parent[a]

  def join(self, a, b):
    a, b = self.find(a), self.find(b)
    if a == b: return

    self.components -= 1
    self.rank[a] += (self.rank[a] == self.rank[b])
    if self.rank[a] < self.rank[b]:
      self.parent[a] = b
    else: self.parent[b] = a

  def __len__(self):
    return self.components
  
  